homeassistant/components/crownstone/entry_manager.py

```diff
@@ -0,0 +1,190 @@
+)
+from .helpers import get_port
+from .listeners import setup_sse_listeners, setup_uart_listeners
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class CrownstoneEntryManager:
+    """Manage a Crownstone config entry."""
+
+    uart: CrownstoneUart
+    cloud: CrownstoneCloud
+    sse: CrownstoneSSEAsync
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+        """Initialize the hub."""
+        self.hass = hass
+        self.config_entry = config_entry
+        self.listeners: dict[str, Any] = {}
+        self.usb_sphere_id: str | None = None
+
+    async def async_setup(self) -> bool:
+        """
+        Set up a Crownstone config entry.
+
+        This class is a combination of Crownstone cloud, Crownstone SSE and Crownstone uart.
+        Returns True if the setup was successful.
+        """
+        email = self.config_entry.data[CONF_EMAIL]
+        password = self.config_entry.data[CONF_PASSWORD]
+
+        self.cloud = CrownstoneCloud(
+            email=email,
+            password=password,
+            clientsession=aiohttp_client.async_get_clientsession(self.hass),
+        )
+        # Login & sync all user data
+        try:
+            await self.cloud.async_initialize()
+        except CrownstoneAuthenticationError as auth_err:
+            _LOGGER.error(
+                "Auth error during login with type: %s and message: %s",
+                auth_err.type,
+                auth_err.message,
+            )
+            return False
+        except CrownstoneUnknownError as unknown_err:
+            _LOGGER.error("Unknown error during login")
+            raise ConfigEntryNotReady from unknown_err
+
+        # A new clientsession is created because the default one does not cleanup on unload
+        self.sse = CrownstoneSSEAsync(
+            email=email,
+            password=password,
+            access_token=self.cloud.access_token,
+            websession=aiohttp_client.async_create_clientsession(self.hass),
+        )
+        # Listen for events in the background, without task tracking
+        asyncio.create_task(self.async_process_events(self.sse))
+        setup_sse_listeners(self)
+
+        # Set up a Crownstone USB only if path exists
+        if self.config_entry.data[CONF_USB_PATH] is not None:
+            await self.async_setup_usb()
+
+        # Save the sphere where the USB is located
+        # Makes HA aware of the Crownstone environment HA is placed in, a user can have multiple
+        self.usb_sphere_id = self.config_entry.data[CONF_USB_SPHERE]
+
+        self.hass.data.setdefault(DOMAIN, {})[self.config_entry.entry_id] = self
+        self.hass.config_entries.async_setup_platforms(self.config_entry, PLATFORMS)
+
+        # HA specific listeners
+        self.config_entry.async_on_unload(
+            self.config_entry.add_update_listener(_async_update_listener)
+        )
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.on_shutdown)
```

Comment:
This needs to be canceled as well if the integration is unloaded or it will leak a listener (wrap in `async_on_unload`)